File: concept/ConceptManagerImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept;

import com.vaticle.typedb.client.api.concept.ConceptManager;
import com.vaticle.typedb.client.api.concept.thing.Attribute;
import com.vaticle.typedb.client.api.concept.thing.Entity;
import com.vaticle.typedb.client.api.concept.thing.Relation;
import com.vaticle.typedb.client.api.concept.type.AttributeType;
import com.vaticle.typedb.client.api.concept.type.EntityType;
import com.vaticle.typedb.client.api.concept.type.RelationType;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.client.common.exception.TypeDBException;
import com.vaticle.typedb.client.concept.thing.AttributeImpl;
import com.vaticle.typedb.client.concept.thing.EntityImpl;
import com.vaticle.typedb.client.concept.thing.RelationImpl;
import com.vaticle.typedb.client.concept.type.AttributeTypeImpl;
import com.vaticle.typedb.client.concept.type.EntityTypeImpl;
import com.vaticle.typedb.client.concept.type.RelationTypeImpl;
import com.vaticle.typeql.lang.common.TypeQLToken;

import javax.annotation.Nullable;
import java.util.List;
import java.util.stream.Collectors;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.MISSING_IID;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.MISSING_LABEL;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_attribute;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_attribute_type;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_entity;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_entity_type;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_relation;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_relation_type;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_get_schema_exceptions;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_put_attribute_type;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_put_entity_type;
import static com.vaticle.typedb.client.jni.typedb_client.concepts_put_relation_type;
import static com.vaticle.typedb.client.jni.typedb_client.schema_exception_code;
import static com.vaticle.typedb.client.jni.typedb_client.schema_exception_message;

public final class ConceptManagerImpl implements ConceptManager {
    final com.vaticle.typedb.client.jni.Transaction nativeTransaction;

    public ConceptManagerImpl(com.vaticle.typedb.client.jni.Transaction nativeTransaction) {
        this.nativeTransaction = nativeTransaction;
    }

    @Override
    public EntityType getRootEntityType() {
        return getEntityType(TypeQLToken.Type.ENTITY.toString());
    }

    @Override
    public RelationType getRootRelationType() {
        return getRelationType(TypeQLToken.Type.RELATION.toString());
    }

    @Override
    public AttributeType getRootAttributeType() {
        return getAttributeType(TypeQLToken.Type.ATTRIBUTE.toString());
    }

    @Override
    @Nullable
    public EntityType getEntityType(String label) {
        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
        com.vaticle.typedb.client.jni.Concept res = concepts_get_entity_type(nativeTransaction, label);
        if (res != null) return new EntityTypeImpl(res);
        else return null;
    }

    @Override
    @Nullable
    public RelationType getRelationType(String label) {
        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
        com.vaticle.typedb.client.jni.Concept res = concepts_get_relation_type(nativeTransaction, label);
        if (res != null) return new RelationTypeImpl(res);
        else return null;
    }

    @Override
    @Nullable
    public AttributeType getAttributeType(String label) {
        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
        com.vaticle.typedb.client.jni.Concept res = concepts_get_attribute_type(nativeTransaction, label);
        if (res != null) return new AttributeTypeImpl(res);
        else return null;
    }

    @Override
    public EntityType putEntityType(String label) {
        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
        return new EntityTypeImpl(concepts_put_entity_type(nativeTransaction, label));
    }

    @Override
    public RelationType putRelationType(String label) {
        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
        return new RelationTypeImpl(concepts_put_relation_type(nativeTransaction, label));
    }

    @Override
    public AttributeType putAttributeType(String label, AttributeType.ValueType valueType) {
        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
        return new AttributeTypeImpl(concepts_put_attribute_type(nativeTransaction, label, valueType.nativeObject));
    }

    @Override
    @Nullable
    public Entity getEntity(String iid) {
        if (iid == null || iid.isEmpty()) throw new TypeDBClientException(MISSING_IID);
        com.vaticle.typedb.client.jni.Concept res = concepts_get_entity(nativeTransaction, iid);
        if (res != null) return new EntityImpl(res);
        else return null;
    }

    @Override
    @Nullable
    public Relation getRelation(String iid) {
        if (iid == null || iid.isEmpty()) throw new TypeDBClientException(MISSING_IID);
        com.vaticle.typedb.client.jni.Concept res = concepts_get_relation(nativeTransaction, iid);
        if (res != null) return new RelationImpl(res);
        else return null;
    }

    @Override
    @Nullable
    public Attribute getAttribute(String iid) {
        if (iid == null || iid.isEmpty()) throw new TypeDBClientException(MISSING_IID);
        com.vaticle.typedb.client.jni.Concept res = concepts_get_attribute(nativeTransaction, iid);
        if (res != null) return new AttributeImpl(res);
        else return null;
    }

    @Override
    public List<TypeDBException> getSchemaExceptions() {
        return concepts_get_schema_exceptions(nativeTransaction).stream()
                .map(e -> new TypeDBException(schema_exception_code(e), schema_exception_message(e)))
                .collect(Collectors.toList());
    }
}

File: concept/ConceptManagerImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { ConceptManager as ConceptProto } from "typedb-protocol/common/concept_pb";
import { Transaction as TransactionProto } from "typedb-protocol/common/transaction_pb";
import { ConceptManager } from "../api/concept/ConceptManager";
import { Thing } from "../api/concept/thing/Thing";
import { AttributeType } from "../api/concept/type/AttributeType";
import { EntityType } from "../api/concept/type/EntityType";
import { RelationType } from "../api/concept/type/RelationType";
import { ThingType } from "../api/concept/type/ThingType";
import { TypeDBTransaction } from "../api/connection/TypeDBTransaction";
import { RequestBuilder } from "../common/rpc/RequestBuilder";
import { AttributeTypeImpl, EntityTypeImpl, RelationTypeImpl, ThingImpl, ThingTypeImpl, } from "../dependencies_internal";
import {Concept} from "../api/concept/Concept";

export class ConceptManagerImpl implements ConceptManager {

    private _transaction: TypeDBTransaction.Extended;

    constructor(client: TypeDBTransaction.Extended) {
        this._transaction = client;
    }

    async getRootThingType(): Promise<ThingType> {
        return this.getThingType("thing");
    }

    async getRootEntityType(): Promise<EntityType> {
        return this.getEntityType("entity");
    }

    async getRootRelationType(): Promise<RelationType> {
        return this.getRelationType("relation");
    }

    async getRootAttributeType(): Promise<AttributeType> {
        return this.getAttributeType("attribute");
    }

    async getThingType(label: string): Promise<ThingType> {
        const request = RequestBuilder.ConceptManager.getThingTypeReq(label);
        const response = await this.execute(request);
        if (response.getGetThingTypeRes().getResCase() == ConceptProto.GetThingType.Res.ResCase.THING_TYPE) {
            return ThingTypeImpl.of(response.getGetThingTypeRes().getThingType());
        } else {
            return null;
        }
    }

    async getEntityType(label: string): Promise<EntityType> {
        const type = await this.getThingType(label);
        if (type?.isEntityType()) return type.asEntityType();
        else return null;
    }

    async getRelationType(label: string): Promise<RelationType> {
        const type = await this.getThingType(label);
        if (type?.isRelationType()) return type.asRelationType();
        else return null;
    }

    async getAttributeType(label: string): Promise<AttributeType> {
        const type = await this.getThingType(label);
        if (type?.isAttributeType()) return type.asAttributeType();
        else return null;
    }

    async getThing(iid: string): Promise<Thing> {
        const request = RequestBuilder.ConceptManager.getThingReq(iid);
        const response = await this.execute(request);
        if (response.getGetThingRes().getResCase() === ConceptProto.GetThing.Res.ResCase.THING) {
            return ThingImpl.of(response.getGetThingRes().getThing());
        } else {
            return null;
        }
    }

    async putEntityType(label: string): Promise<EntityType> {
        const request = RequestBuilder.ConceptManager.putEntityTypeReq(label);
        const response = await this.execute(request);
        return EntityTypeImpl.of(response.getPutEntityTypeRes().getEntityType());
    }

    async putRelationType(label: string): Promise<RelationType> {
        const request = RequestBuilder.ConceptManager.putRelationTypeReq(label);
        const response = await this.execute(request);
        return RelationTypeImpl.of(response.getPutRelationTypeRes().getRelationType());
    }

    async putAttributeType(label: string, valueType: Concept.ValueType): Promise<AttributeType> {
        const request = RequestBuilder.ConceptManager.putAttributeTypeReq(label, valueType.proto());
        const response = await this.execute(request);
        return AttributeTypeImpl.of(response.getPutAttributeTypeRes().getAttributeType());
    }

    private execute(request: TransactionProto.Req): Promise<ConceptProto.Res> {
        return this._transaction.rpcExecute(request).then((res) => res.getConceptManagerRes());
    }

}
